glyphs-reader/src/glyphdata.rs

```diff
@@ -85,11 +126,170 @@ impl GlyphData {
     }
 
     /// Look up info for a glyph by codepoint
-    pub fn get_by_codepoint(&self, codepoint: u32) -> Option<&GlyphInfo> {
+    fn get_by_codepoint(&self, codepoint: u32) -> Option<&GlyphInfo> {
         self.unicode_map
             .get(&codepoint)
             .and_then(|idx| self.data.get(*idx as usize))
     }
+
+    // https://github.com/googlefonts/glyphsLib/blob/e2ebf5b517d/Lib/glyphsLib/glyphdata.py#L199
+    fn construct_category(&self, name: &str) -> Option<(Category, Subcategory)> {
+        if name.starts_with('_') {
+            return None;
+        }
+        let base_name = self
+            .split_glyph_name(name)
+            .map(|(base, _)| base)
+            .unwrap_or(name);
+        if let Some(info) = self.get_by_name(base_name) {
+            return Some((info.category, info.subcategory));
+        }
+
+        if let Some(base_names) = self.split_ligature_glyph_name(base_name) {
```

Comment:
This block is a touch hard to follow at a glance, a wee comment explaining what's happening would be welcome.

Seems almost worthy of it's own function.

Indeed ... I wonder if several of these blocks could be functions returning Option and this whole function becomes an unwrap_or_else chain?

Reply:
a lot of this is about me wanting to try and fairly closely follow the source I'm working from.

I tried refactoring this block into a separate function but then the overall control flow feels harder to read, because you need to jump down to read that block before continuing. The bit below can go into a separate function, though!